graph/src/blockchain/substreams_block_stream.rs

```diff
@@ -0,0 +1,319 @@
+                                    None => {}
+                                    Some(BlockResponse::Proceed(event, cursor)) => {
+                                        // Reset backoff because we got a good value from the stream
+                                        backoff.reset();
+
+                                        metrics.observe_response("proceed", &mut last_response_time);
+
+                                        yield event;
+
+                                        latest_cursor = cursor;
+                                    }
+                                }
+                            },
+                            Err(err) => {
+                                info!(&logger, "received err");
+                                // We have an open connection but there was an error processing the Firehose
+                                // response. We will reconnect the stream after this; this is the case where
+                                // we actually _want_ to back off in case we keep running into the same error.
+                                // An example of this situation is if we get invalid block or transaction data
+                                // that cannot be decoded properly.
+
+                                metrics.observe_response("error", &mut last_response_time);
+
+                                error!(logger, "{:#}", err);
+                                expected_stream_end = true;
+                                break;
+                            }
+                        }
+                    }
+
+                    if !expected_stream_end {
+                        error!(logger, "Stream blocks complete unexpectedly, expecting stream to always stream blocks");
+                    }
+                },
+                Err(e) => {
+                    // We failed to connect and will try again; this is another
+                    // case where we actually _want_ to back off in case we keep
+                    // having connection errors.
+
+                    metrics.observe_failed_connection(&mut connect_start);
+
+                    error!(logger, "Unable to connect to endpoint: {:?}", e);
+                }
+            }
+
+            // If we reach this point, we must wait a bit before retrying, unless `skip_backoff` is true
+            if !skip_backoff {
+                backoff.sleep_async().await;
+            }
+        }
+    }
+}
+
+enum BlockResponse<C: Blockchain> {
+    Proceed(BlockStreamEvent<C>, String),
+}
+
+async fn process_substreams_response<C: Blockchain, F: SubstreamsMapper<C>>(
+    result: Result<Response, Status>,
+    mapper: &F,
+    logger: &Logger,
+) -> Result<Option<BlockResponse<C>>, Error> {
+    let response = match result {
+        Ok(v) => v,
+        Err(e) => return Err(anyhow!("An error occurred while streaming blocks: {:?}", e)),
+    };
+
+    match response.message.unwrap() {
```

Comment:
Could this just handle the error or is there a reason this is safe? I think I would rather handle this explicitly rather than using unwrap unless this intentional, in which case can you please add a comment?